src/main/bootstrap/cache_validator.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::common::{ipv4_addr, PeerInfo, State};
+use crate::main::{CrustData, EventLoopCore};
+use crate::nat::GetExtAddr;
+use mio::{Poll, PollOpt, Ready, Token};
+use mio_extras::channel::{channel, Receiver, Sender};
+use safe_crypto::{PublicEncryptKey, SecretEncryptKey};
+use std::any::Any;
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::net::SocketAddr;
+use std::rc::{Rc, Weak};
+
+/// When peer is tested, if it's still online, we send an encrypted request and wait for response.
+/// This is the maximum time we wait before declaring that peer is dead.
+const PEER_TEST_TIMEOUT_SEC: u64 = 30;
+
+/// Bootstrap cache validator tests inactive cached peers if they are still alive.
+/// It receives peers from async in-memory channel.
+/// Cache validator is a future/state that never finishes.
+pub struct CacheValidator {
+    token: Token,
+    our_pk: PublicEncryptKey,
+    our_sk: SecretEncryptKey,
+    expired_rx: Receiver<HashSet<PeerInfo>>,
+    expired_tx: Sender<HashSet<PeerInfo>>,
+    /// Tokens for in progress states testing peer connectivity.
+    sent_requests: HashSet<Token>,
+    /// When need to reference self from the callbacks.
+    self_weak: Weak<RefCell<CacheValidator>>,
+}
+
+impl CacheValidator {
+    /// Starts running bootstrap cache entry validation.
+    pub fn start(
+        core: &mut EventLoopCore,
+        poll: &Poll,
+        our_pk: PublicEncryptKey,
+        our_sk: SecretEncryptKey,
+    ) -> crate::Res<()> {
+        let token = core.get_new_token();
+
+        let (expired_tx, expired_rx) = channel();
```

Comment:
I'm not liking this mio::channel to communicate between states all residing intra-thread. Channels are more for cross threaded communication so I'm thinking we can have a different design but need to jump somewhere else just now - will come back to this. Basically it feels like an overkill - like using Arc's and Mutexes and Atomics for intra-thread communication with needless overhead. However this might be the best compromise but will need to come back to this to see if there's no alternate/better way.

Reply:
Mio async channel was used for notifications. Under the hood it should be just an in-mem buffer with some notification mechanism. So it shouldn't introduce more overhead compare to other solutions, cause some sort of buffering will be used anyway.

Reply:
What about instead of having a sender to the cache-validator kept as user-data in `Core` and using that to go via Poll indirectly, have `CacheValidator` itself (as an `Rc<RefCell` which is what it is here anyway) lodged there ? So now we can have direct invocations of `CacheValidator` facilities which prevents going through a channel route. This is certainly less overhead than if we used channels is it not ? Considering the channels would have to do additional work including thread safety guarantees (when we don't need it) and posting to the Poll loop.